### PR TITLE
Compensate for body tag with position absolute, relative or static

### PR DESCRIPTION
**Pull request by jasco**
In cases where the body tag has a position style of absolute, relative or static adjusts the positioning to compensate for position offsets and borders so powerTip remains properly placed.

Also opts for making all placement based on top rather than bottom attribute.  The placement can be adjusted to use bottom if there is some reason to prefer it.  The tests have also been adjusted.

This extra flexibility was needed because the control is used in pages where I do not control the html.

Comments or suggestions for improvement welcomed.

### CODE REVIEW

**Comment by stevenbenner** [comment]
This looks real good, and would be a huge improvement. I need to do a detailed code review. I do have a concern from the first-pass:

* **Tips that were using `bottom` for CSS position now use `top`.**

	The reason that they were using `bottom` was because it is more reliable. If we use `bottom` for the northern tooltips then if the content in the tooltip changes shape the tooltip will remain in the correct position. (See: #31 for this original disucssion)

Would it be possible to refactor this to keep using the `bottom` CSS rule on north-* tooltips?

**Comment by jasco** [comment]
@stevenbenner The positioning has been updated to use bottom for north placements.  The branch has also been rebased to current tip of master.

**Comment by jonwolfe** [comment]
@stevenbenner thanks for the great library. We've incorporated PowerTip into a new project, and are experiencing this issue.

As we decide whether to implement a workaround or do a temporary monkeypatch of the current PowerTip release, it would be great to know if you plan on accepting this PR.

Thanks again.

**Comment by stevenbenner** [comment]
@jonwolfe Yes, I do plan on accepting this pull request. However it won't be in the next version (1.3.0). It's so close to release, and this change has more risk than I'm comfortable with for 1.3.0. Hopefully it will get in for version 1.3.1.

**Comment by jasco** [comment]
As an aside, perhaps higher risk commits fit better in a major revision release rather than a minor one?

In any event, it turns out that some webpages position the html tag rather then the body tag.  I have a reworked patch that checks both body and html.  Would you prefer that I submit a separate pull request or update this one?

**Comment by stevenbenner** [comment]
@jasco How big is the change? If it's a significant rewrite of this patch then go ahead and open a new pull request. Might be educational for me to see the changes. If it's just tweaking a couple lines then go ahead and update this pull request.

**Comment by jasco** [comment]
The technique is the same but the compensation calculation code was significantly restructured. In order to check body and html tags in turn the CSS position check and offset calculation were decomposed into functions. Upon further consideration I think it makes sense to update this patch since the updated implementation really does supersede this one because the checks are ordered.  I will keep the commit history so if you can see both versions.

Since I am not sure the issue was clearly articulated before, I'll summarize. The issue is that absolute positioning is relative to the closest, non-static ancestor.  Only when all ancestors are static is it relative to the viewport.  The compensation adjusts the coordinates relative to viewport to the coordinates relative to the anchoring element.

File: src/utility.js [hunk 1]
```diff
@@ -200,3 +204,54 @@ function countFlags(value) {
 	}
 	return count;
 }
+
+/**
+ * Compute compensating position offsets if body or html element has non-static position attribute.
+ * @private
+ * @param {number} windowWidth Window width in pixels.
+ * @param {number} windowHeight Window height in pixels.
+ * @return {Offsets} The top, left, right, bottom offset in pixels
```
Comment (stevenbenner) [inline src/utility.js]: "Offsets" isn't a type in this context. This should be "Object". Example:

`@return {Object} An object with top, left, right, bottom offset values`

File: src/utility.js [hunk 2]
```diff
@@ -200,3 +204,54 @@ function countFlags(value) {
...
+ */
+function computePositionCompensation(windowWidth, windowHeight) {
+	// Check if the element is "positioned". A "positioned" element has a CSS
+	// position value other than static. Whether the element is positioned is
+	// relevant because absolutely positioned elements are positioned relative
+	// to the first positioned ancestor rather than relative to the doc origin.
+	var isPositioned = function(el) {
```
Comment (stevenbenner) [inline src/utility.js]: I'd like to see this function pulled out of the `computePositionCompensation()` function. This has a couple benefits:

1. (Big one) It can be unit tested on it's own.
2. It makes sure that the function is only generated once at run time.

Just move it to be another utility function in this file.

File: src/utility.js [hunk 3]
```diff
@@ -200,3 +204,54 @@ function countFlags(value) {
...
+		return el.css('position') !== 'static';
+	};
+
+	var getElementOffsets = function(el) {
```
Comment (stevenbenner) [inline src/utility.js]: x2 - Same thing as above. I'd like this function to be moved out of `computePositionCompensation()` and be a global utility function.

File: src/utility.js [hunk 4]
```diff
@@ -200,3 +204,54 @@ function countFlags(value) {
...
+		elPositionPx = el.position();
+
+		// Compute the far margins based off the outerWidth values.
+		elWidthWithMargin = el.outerWidth(true);
+		elHeightWithMargin = el.outerHeight(true);
+
+		// right offset = right margin + body right position
+		offsets.right = (elWidthWithMargin - el.outerWidth() - (offsets.left - elPositionPx.left)) + (windowWidth - elWidthWithMargin - elPositionPx.left);
+		// bottom offset = bottom margin + body bottom position
+		offsets.bottom = (elHeightWithMargin - el.outerHeight() - offsets.top) + (windowHeight - elHeightWithMargin - elPositionPx.top);
```
Comment (stevenbenner) [inline src/utility.js]: Are the `elPositionPx` values actually needed? They seems to be causing a bug (I'll detail in PR comments).

**Review by stevenbenner — CHANGES_REQUESTED** [review]
This looks pretty good. I found one bug that I'd like to get your feedback on, and made a couple other little notes.

One other thing I'd like to change is the extra test pages. These are probably not necessary. I'd rather modify the current edge-cases.html file to include a css toggler. That way I can review the whole suite of extra test cases with the new position handling.

I can do this later if you don't want to go through the trouble. The idea would be to add a select box, radio, or button that would then set or remove the CSS `position` on the `body` and/or `html` tags.

**Comment by stevenbenner** [comment]
I've had a chance to test this out and do a code review, and I did find one bug.

#### Normal
normal

#### bodyoffset-rel
bottom-gap
_Env: Firefox 51.0.1_

Notice that there is a gap between the tooltip arrow and the element the tooltip opened for on the relative positioned test. I think this is because you're adding the `$.postion()` value to the returned offsets. I don't see why that is needed, is there a reason you're using `elPositionPx`?

**Comment by jasco** [comment]
Good catch, thank you.  Not sure why I did not notice the offset.  It seems that when calculated as I intended that the el*WithMargin and the elPositionPx terms all cancel out.  The simplified calculation seems to fix the anomaly.

**Comment by jasco** [comment]
I added the position settings to edge-cases.

It works fine for body but not for the html tag.  The jquery offset returns left: 0, top: 0 even when the html tag is positioned.  I'm not sure the most robust way to find the position yet.  getBoundingClientRect has some potential but I do not yet have the right computation for right and bottom offsets.

I'll push the tests.  Maybe you have a suggestion?

File: src/utility.js [hunk 5]
```diff
@@ -165,8 +167,10 @@ function getTooltipContent(element) {
  * @return {number} Value with the collision flags.
  */
 function getViewportCollisions(coords, elementWidth, elementHeight) {
-	var viewportTop = session.scrollTop,
-		viewportLeft =  session.scrollLeft,
+	// adjusting viewport even though it might be negative because coords
+	// comparing with are relative to compensated position
+	var viewportTop = session.scrollTop - session.positionCompensation.top,
+		viewportLeft =	session.scrollLeft - session.positionCompensation.left,
```
Comment (stevenbenner) [inline src/utility.js]: These seem to be causing the collision detection to break when the position is relative. Will explain the bug in PR comments.

Are you sure this is the correct place to apply the adjustment? These variables (`viewportTop`, `viewportLeft`) are the viewport position relative to the page.

desktop_page
_(Source: http://www.quirksmode.org/mobile/viewports.html)_


Since the `positionCompensation` adjustments are based on the CSS, I think the adjustments need to be applied in the maths below, which compare the `CSSCoordinates` values to the viewport position to determine if there is a collision.

File: test/edge-cases.html [hunk 6]
```diff
@@ -29,11 +29,54 @@
...
+				window.dispatchEvent(new Event('resize'));
+			}
+
 			// run theme switcher on page load
 			setTheme();
 
-			// hook theme switcher to select change
+			// hook theme and position switcher to select change
 			$('#theme-switcher').on('change', setTheme);
+			$('#position-switcher').on('change', setCssPosition);
```
Comment (stevenbenner) [inline test/edge-cases.html]: Could you also call this function on load (like `setTheme()` on line 75)? It's nice to be able to reload the page and not have to scroll to the top to reset the position.

Just add `setCssPosition()` somewhere.

**Review by stevenbenner — CHANGES_REQUESTED** [review]
This latest version has fixed the bottom placement bug in my testing. Though I did discover one other bug.

Thanks for adding the positioning tool to tests-edge.html. That is really convenient for my testing.

Since all of the testing can now be done on one page, these files are no longer needed:
* test/htmloffset-rel.html
* test/bodyoffset-rel.html
* test/bodyoffset-abs.html
* test/tests-bodyoffset.js

**Comment by stevenbenner** [comment]
Thanks for the updates! The latest version fixed that bug, though I did discover another bug in testing.

### Smart placement broken in relative positioning

It looks like the smart placement feature isn't working properly for tooltips that use `bottom` or `right` css positioning when the page is using relative positioning.

**Repro steps:**
1. Open tests-edge.html
2. Set the position to "BODY relative"
3. Scroll down to the "Huge Text with Smart Placement" test
4. Mouse over the "North" button

You'll see that even if there is plenty of room the tooltip will not show in the north position.

I believe this is because the collision detection from the `getViewportCollisions()` function is broken. This happens when the `positionCompensation.top` value is large. After reviewing the code I think that the `positionCompensation` adjustments are being applied in the wrong place (see my comment in the code review).

### Side thought

After reviewing the code I've started to think that maybe this new logic belongs in the `CSSCoordinates()` object. It might be convenient to change the `top`, `left`, `right`, and `bottom` properties to be methods/getters that automatically apply the `positionCompensation` adjustments when something tries to read the property. That way you wouldn't have to apply the compensation to every place in code that touches the CSS positioning values.

This is just a thought I had, not a change request. So feel free to ignore.

**Comment by jasco** [comment]
(Rebased to current master.  It appeared there was no conflict between this patch and the diff to master.)

Thanks for your review and feedback on this issue.  I be out the rest of the week but I'll look at your comments when I get back.

File: test/edge-cases.html (hunk 6 above)
Comment (jasco) [inline test/edge-cases.html]: I would think it will need more than the function call.  The dropdown value would presumably reset on reload unless the value is saved somewhere.  I could use a url parameter to preserve the position setting so it can be restored on reload.  Would that be helpful?

**Comment by jasco** [comment]
The compensation calculations and collision detection have been completely revamped.  I also took your suggestion to move the compensation computation into CSSCoordinates.  The collision detection has been revised to use the same origin for all values include bottom and right.  Also, the calculations now accounts for non-zero border widths on the non-static ancestor.

Also, the branch has been rebased to the head of master again.

I hope/think it is right now.  I finally feel like I completely understand how measurements work in all of the cases I considered.  It has been a rather painful process but I learned something along the way.  I would note that, while I don't know of any issues, I have not reviewed for browser compatibility or behavior in legacy doctypes.
